gadgets/trace_ssl/program.bpf.c

```diff
@@ -0,0 +1,432 @@
+	__uint(max_entries, MAX_ENTRIES);
+	__type(key, u32); // tid
+	__type(value, u64); // timestamp
+} ssl_starttime SEC(".maps");
+
+struct {
+	__uint(type, BPF_MAP_TYPE_HASH);
+	__uint(max_entries, MAX_ENTRIES);
+	__type(key, u32); // tid
+	__type(value, void *); // buffer
+} ssl_bufs SEC(".maps");
+
+/* used for context between uprobes and uretprobes of libcrypto operations */
+struct {
+	__uint(type, BPF_MAP_TYPE_HASH);
+	__uint(max_entries, MAX_ENTRIES);
+	__type(key, u32); // tid
+	__type(value, u64); // timestamp
+} crypto_starttime SEC(".maps");
+
+/**
+ * clean up the maps when a thread terminates,
+ * because there may be residual data in the map
+ * if a userspace thread is killed between a uprobe and a uretprobe
+ */
+SEC("tracepoint/sched/sched_process_exit")
+int trace_sched_process_exit(void *ctx)
+{
+	u32 tid;
+
+	tid = (u32)bpf_get_current_pid_tgid();
+	bpf_map_delete_elem(&mntns_context, &tid);
+	bpf_map_delete_elem(&ssl_starttime, &tid);
+	bpf_map_delete_elem(&ssl_bufs, &tid);
+	bpf_map_delete_elem(&crypto_starttime, &tid);
+	return 0;
+}
+
+static __always_inline int probe_ssl_rw_enter(struct pt_regs *ctx, void *buf)
+{
+	u64 mntns_id;
+	u32 tid;
+	u64 ts;
+
+	mntns_id = gadget_get_mntns_id();
+	if (gadget_should_discard_mntns_id(mntns_id))
+		return 0;
+
+	tid = (u32)bpf_get_current_pid_tgid();
+	ts = bpf_ktime_get_boot_ns();
+
+	bpf_map_update_elem(&mntns_context, &tid, &mntns_id, BPF_ANY);
+	bpf_map_update_elem(&ssl_bufs, &tid, (u64 *)&buf, BPF_ANY);
+	bpf_map_update_elem(&ssl_starttime, &tid, &ts, BPF_ANY);
+	return 0;
+}
+
+static __always_inline int probe_ssl_rw_exit(struct pt_regs *ctx,
+					     enum operation op)
+{
+	u64 *mntns_ptr;
+	int len;
+	u64 *bufp;
+	u64 *tsp;
+	u32 buf_copy_size;
+	struct event *event;
+	u64 pid_tgid;
+	u64 uid_gid;
+	u32 pid;
+	u32 tid;
+	u64 ts;
+
+	pid_tgid = bpf_get_current_pid_tgid();
+	uid_gid = bpf_get_current_uid_gid();
+	pid = pid_tgid >> 32;
+	tid = (u32)pid_tgid;
+	ts = bpf_ktime_get_boot_ns();
+
+	len = PT_REGS_RC(ctx);
+	if (len <= 0) // no data
+		goto clean;
+
+	bufp = bpf_map_lookup_elem(&ssl_bufs, &tid);
+	if (bufp == 0)
+		goto clean;
+	bpf_map_delete_elem(&ssl_bufs, &tid);
+
+	tsp = bpf_map_lookup_elem(&ssl_starttime, &tid);
+	if (tsp == 0)
+		goto clean;
+
+	mntns_ptr = bpf_map_lookup_elem(&mntns_context, &tid);
+	if (mntns_ptr == 0)
+		goto clean;
+
+	buf_copy_size = (size_t)MAX_BUF_SIZE < (size_t)len ?
+				(size_t)MAX_BUF_SIZE :
+				(size_t)len;
+	event = gadget_reserve_buf(&events, sizeof(struct event));
+	if (!event)
+		goto clean;
+
+	event->mntns_id = *mntns_ptr;
+	event->operation = op;
+	event->timestamp = ts;
+	event->latency_ns = ts - *tsp;
+	event->pid = pid;
+	event->tid = tid;
+	event->uid = uid_gid;
+	event->gid = uid_gid >> 32;
+	event->len = len;
+	event->retval = PT_REGS_RC(ctx);
+
+	bpf_get_current_comm(&event->comm, sizeof(event->comm));
+
+	if (!record_data ||
+	    bpf_probe_read_user(&event->buf, buf_copy_size, (char *)*bufp))
```

Comment:
I get the following:

> Error: starting operator "oci": starting operator "ebpf": creating eBPF collection: program trace_uretprobe_libssl_SSL_write: load program: permission denied: 102: (85) call bpf_probe_read_user#112: R2 unbounded memory access, use 'var &= const' or 'if (var < const)' (133 line(s) omitted)

Reply:
Got the same err message in 6.6, seems it's because the low version of verifier.

Reply:
Although `buf_copy_size` can be up to `MAX_BUF_SIZE`, and using `MAX_BUF_SIZE - 1` will waste 1 byte, I haven't found a better way to make the verifier happy. Let's go with this.